lldb/source/Target/SwiftLanguageRuntime.cpp

```diff
@@ -220,59 +239,95 @@ AppleObjCRuntimeV2 *SwiftLanguageRuntime::GetObjCRuntime() {
+llvm::Optional<lldb::addr_t> SwiftLanguageRuntime::GetSwiftNativeNSErrorISA() {
+  if (!m_initialized_swift_native_error_isa)
+    SetupSwiftError();
+  assert(m_initialized_swift_native_error_isa);
+  return m_SwiftNativeNSErrorISA;
 }
 
 void SwiftLanguageRuntime::SetupExclusivity() {
-  Target &target(m_process->GetTarget());
-
-  ConstString g_disableExclusivityChecking("_swift_disableExclusivityChecking");
-
   m_dynamic_exclusivity_flag_addr = FindSymbolForSwiftObject(
-      target, g_disableExclusivityChecking, eSymbolTypeData);
+      m_process->GetTarget(), RuntimeKind::Swift,
+      "_swift_disableExclusivityChecking", eSymbolTypeData);
+  m_initialized_dynamic_exclusivity_flag_addr = true;
 
   Log *log(GetLogIfAnyCategoriesSet(LIBLLDB_LOG_EXPRESSIONS));
-
   if (log)
     log->Printf("SwiftLanguageRuntime: _swift_disableExclusivityChecking = %lu",
                 m_dynamic_exclusivity_flag_addr ?
                 *m_dynamic_exclusivity_flag_addr : 0);
 }
 
+llvm::Optional<lldb::addr_t>
+SwiftLanguageRuntime::GetDynamicExclusivityFlagAddr() {
+  if (!m_initialized_dynamic_exclusivity_flag_addr)
+    SetupExclusivity();
+  assert(m_initialized_dynamic_exclusivity_flag_addr);
+  return m_dynamic_exclusivity_flag_addr;
+}
+
 void SwiftLanguageRuntime::SetupABIBit() {
+  if (m_initialized_swift_classIsSwiftMask)
+    return;
   Target &target(m_process->GetTarget());
-  ConstString g_objc_debug_swift_stable_abi_bit("objc_debug_swift_stable_abi_bit");
-
-  if (FindSymbolForSwiftObject(target, g_objc_debug_swift_stable_abi_bit, eSymbolTypeAny))
+  if (FindSymbolForSwiftObject(target, RuntimeKind::ObjC,
+                               "objc_debug_swift_stable_abi_bit",
+                               eSymbolTypeAny))
     _swift_classIsSwiftMask = 2;
   else
     _swift_classIsSwiftMask = 1;
+
+  m_initialized_swift_classIsSwiftMask = true;
 }
 
-void SwiftLanguageRuntime::ModulesDidLoad(const ModuleList &module_list) {
-  module_list.ForEach([&](const ModuleSP &module_sp) -> bool {
+bool SwiftLanguageRuntime::AddModuleToReflectionContext(
+    const lldb::ModuleSP &module_sp) {
+  // This function is called from within SetupReflection so it cannot
+  // call GetReflectionContext().
+  assert(m_initialized_reflection_ctx);
+  if (!m_reflection_ctx)
+    return false;
+  if (!module_sp)
+    return false;
   auto *obj_file = module_sp->GetObjectFile();
-    if (!obj_file)
-        return true;
-    Address start_address = obj_file->GetBaseAddress();
-    auto load_ptr = static_cast<uintptr_t>(
-        start_address.GetLoadAddress(&(m_process->GetTarget())));
-    if (load_ptr == 0 || load_ptr == LLDB_INVALID_ADDRESS)
-      return false;
-    if (!reflection_ctx)
-      return false;
-    if (HasReflectionInfo(obj_file))
-      reflection_ctx->addImage(swift::remote::RemoteAddress(load_ptr));
-    return true;
-  });
+  if (!obj_file)
+    return false;
+  if (obj_file->GetPluginName().GetStringRef().equals("elf"))
+    return false;
```

Comment:
This is not new with this patch, but does this mean we just are not able to do any RemoteMirrors queries on ELF platforms?

Reply:
Correct. I believe that in their current state they were prone to crashing. @dcci knows the ugly details.